### DISCUSSION TITLE
Performance Test DuckDB 0.7.1 vs H2 2.2.214, what am I doing wrong?

### DISCUSSION THREAD

**Discussion by manticore-projects** [duckdb/duckdb#7389] [Q&A]
Greetings!

I would like to ask for some help please since I get very baffling results from my performance test.
I want to dice-up a large table which has columns like this:

Unique ID | key1 | key2 | key3 | key4 | BigDecimal value1 | BigDecimal value2 | BigDecimal value3 | ... | BigDecimal valueN

<!DOCTYPE html>

<html>

<body>

ID_INSTRUMENT | PRODUCT | SEGMENT | TYPE | CURRENCY | REMAINING_PRINCIPAL | AMORTISED_COST | EIR
-- | -- | -- | -- | -- | -- | -- | --
1 | a | reguler | syariah | IDR | -2248872.98 | -2255306.04 | 0.239228891069
2 | a | reguler | syariah | IDR | -2065395.52 | -2071246.38 | 0.239156737371
3 | a | reguler | syariah | IDR | -2169132.6 | -2175313.11 | 0.239199137356
4 | a | reguler | konvensional | IDR | -1879831.6 | -1885088.98 | 0.239068194532
5 | a | reguler | syariah | IDR | -2009886.19 | -2015560.06 | 0.239132106932
6 | a | reguler | konvensional | IDR | -1791737.53 | -1796711.1 | 0.239019128002
7 | a | reguler | konvensional | IDR | -1828718.13 | -1833811.02 | 0.239040358855
8 | a | reguler | konvensional | IDR | -1773189.6 | -1778103.21 | 0.2390081263
9 | a | reguler | syariah | IDR | -1889837.57 | -1895127.1 | 0.239073444807
10 | a | reguler | syariah | IDR | -2128432.39 | -2134483.74 | 0.239183032709
11 | a | reguler | syariah | IDR | -2098910.17 | -2104867.68 | 0.239170928973
</body>
</html>

The goal was to get a better performance, when querying any Total of the values, e. g.
```sql
select key1, key2, sum(value1), sum(value2)
from testTable
group by key1, key2
```

So I have set up a comprehensive benchmark which:
1) Define a table `testTable` in a H2 and DuckDB memory database, index the Key columns
2) Fill the table in the H2 database with a randomized Sample of 1M records, using 4 different key columns and 2 different key values for each key column. Furthermore the BigDecimal value columns are properly randomized. 
3) extract the H2 Table into a Parquet File
4) import the Parquet File into the DuckDB

5) execute the query on H2 and measure the time (using `JMH`)
6) execute the query on DuckDB and measure the time (using `JMH`)

I would have expected, that DuckDB would perform much faster than H2 since in my understanding this is an OLAP/Analytics scenario (and not a transactional scenario).

However, the results shocked me and I think that I did a mistake somewhere or have a misunderstanding: 
```text
Benchmark                   Mode  Cnt   Score   Error  Units
PerformanceTest.queryH2     avgt    2   0.033          ms/op
PerformanceTest.testDuckDB  avgt    2  24.586          ms/op
```

[Code is here](https://github.com/manticore-projects/JDBCParquetWriter), you can simply run it via
```shell
gradle jmh
```

Can anyone explain to me please, what I am doing wrong and why DuckDB is not faster than H2 in my scenario?

**Comment by manticore-projects** [comment]
I figured out how to disable the H2 Query Cache and this gives more credible results which seem technically correct.
```java
DriverManager.getConnection("jdbc:h2:mem:;QUERY_CACHE_SIZE=0");
```

```text
Benchmark                   Mode  Cnt    Score    Error  Units
PerformanceTest.queryH2     avgt    3  289.737 ± 79.437  ms/op
PerformanceTest.testDuckDB  avgt    3  293.560 ±  7.939  ms/op
```

So I can ask now the more interesting question: Why is DuckDB not faster than H2 despite solving an OLAP/Analytics query?
I would have expected DuckDB to excel especially for this kind of use case (while falling back when manipulating single rows in a table).

Please help me to understand this issue.

**Reply by Alex-Monahan** [reply]
Howdy! Thanks for experimenting with DuckDB!

Could you rerun with a larger dataset? Maybe 100x larger? 1 million rows is still fairly small.

Could you also provide your entire script? Maybe you are measuing connection time in addition to execution time?

Thanks!

**Reply by manticore-projects** [reply]
Thank you for the feedback.

Yes, will do -- although I may have to switch to a disk storage for those volumes.
The 1 Mill rows occupy 110 MB memory -- I don't have 100k times of that :(

> Could you also provide your entire script? Maybe you are measuing connection time in addition to execution time?
Self containing test is there: https://github.com/manticore-projects/JDBCParquetWriter
You can run it via `gradle jmh`

The particular script is here: https://github.com/manticore-projects/JDBCParquetWriter/blob/main/src/test/java/com/manticore/jdbc/parquetwriter/PerformanceTest.java

**Reply by manticore-projects** [reply]
One more interesting observation: The test runs fine with DuckDB-0.7.1 but hangs forever with 0.8.0-268 Snapshot.
No idea what's going on, I just changed the JAR.

**Comment by Mytherin** [comment]
How many rows are in the result? If you are reading a large result set back into Java the time is likely dominated by the result conversion - not the actual query.

**Reply by manticore-projects** [reply]
Thank you for responding.
I know what you mean and we get only a very small ResultSet of 6 records, since we fully aggregate over all rows:

```sql
select key1, key2, sum(value1), sum(value2)
from testTable
group by key1, key2
```

The work is on the DB, not on the Result Set Conversion.

**Comment by manticore-projects** [comment]
Good news: When increasing the Volume above 1M records, DuckDB starts to shine.

```text
Benchmark                   Mode  Cnt      Score       Error  Units
PerformanceTest.queryH2     avgt    3  18897.428 ± 59760.573  ms/op
PerformanceTest.testDuckDB  avgt    3  12176.611 ±  3547.932  ms/op
```

**Comment by l1t1** [comment]
10M rows:
```
create table t as select i::int i,(i*2)::int j,(random()*i)::int%3 a,(random()*i)::int%4 b from range(1,1e7::int)t(i);

select a,b,sum(i),sum(j) from t group by a,b;
```

**Reply by manticore-projects** [reply]
Thank you for pointing this out!
However, my objective is focusing on: a) extracting a Parquet File from a Java ResultSet and b) testing dicing/OLAP performance for my particular, financial data use case. (Since my data will always start "transactional" first and become "OLAP" only at the very end for financial transactions).

Especially the **KEY columns need to be populated with a certain distribution weight**, not just randomly. (In my example, I need 84% in Stage 1, 14% in Stage 2 and 2% in Stage 3). That is why I have to go the long and complicated way.

I will use your query as challenge for JSQLParser though and implement the support for this JSON Table now, so thanks again.

**Comment by ryanhamilton** [comment]
Hi Manticore,

It seems we are working on very similar problems resultsets, parquet, conversion, using H2 or duckdb as cache. 
Here is the insert benchmark I ran on both duckdb and H2 that you may find useful: https://www.timestored.com/data/duckdb/insert-benchmark.

I make a free SQL tool called QStudio, within it you can save any queried data locally to a duckdb database as parquet. This involves converting resultsets from anywhere (clickhouse, postgresql, ms server) to parquet.  Would be good to chat on our use-cases, might be a way to work together. QStudio is open source.
